### PR TITLE
Proposal: Remove Composer's Vendor-Directory from Source-Tree

### PR DESCRIPTION
**Pull request by klees**
Dear All!

I hereby propose to remove the directory `libs\composer\vendor`, created by composer, from our source-tree.

I understand that this is a departure from the current principle that ILIAS should be installable directly from the repo. The current situation that the vendor directory is included in our source-tree, still causes a lot of friction, namely:
* In my local development environment I need to install dev-dependencies to perform automated tests. This leads to changes in several files that git will display besides the real changes that I made. The mass of these changes often trumps my real changes. I need to be careful to pick the right changes for my commit. This is especially tedious when something happens in the dependencies.
* Similar for PRs: Diffs in PRs contain unintelligible, unreadable or uninteresting parts when libraries or classes are added to or removed from the repo. This is impressively proved by this very PR.
* People forget to update the classmap when it needs to be updated. ILIAS then might be installable but does not function properly under all intricate circumstances.
* People think that directly modifying the dependencies is fine. Or they modify the dependencies by accident, e.g. during a merge.

To maintain a repo as a source for installations, we could teach travis to update a second pull-only repo regularly, e.g. every night or week, or on every new release-tag. If this move here is supported I'll be happy to implement that automatism.

Finally: There are complaints in the community about missing resources regarding development. This move will free some of the existing resources in that field and will allow us to increase the developer experience further.

Best regards!

### CODE REVIEW

**Comment by klees** [comment]
lol

Bildschirmfoto von 2019-05-10 09-33-56

File: libs/composer/README.md [hunk 1]
```diff
@@ -62,16 +62,27 @@ composer remove --dev <DEPENDENCY_NAME>
...
 ```bash
-$ composer dump-autoload
+composer install --no-dev
+composer dump-autoload
 ```
 
+To populate the vendor-directory for a development environment, use
+
+```bash
+composer install --no-dev
```
Comment (kergomard) [inline libs/composer/README.md:80]: This should be:
```suggestion
composer install 
```
right?

**Comment by matthiaskunkel** [comment]
Jour Fixe, 20 MAY 2019: We highly appreciate this PR and accept it for trunk. We do not see a need for a second respository. And we need to consider the changed process in the build process for ILIAS releases. There should also be a simple solution for admins to check if their composer lib is up-to-date or not (similar to missing DB update notification).

**Comment by mjansenDatabay** [comment]
:shipit:

**Comment by klees** [comment]
I gave this

> There should also be a simple solution for admins to check if their composer lib is up-to-date or not (similar to missing DB update notification).

some thought.

First thing to consider is, if we would want to include composer as a production dependency. I would not do that. So we cannot use composer (e.g. [`composer install`](https://getcomposer.org/doc/03-cli.md#install-i)) or a wrapper around it to reliably alarm admins of a production system or perform updates automatically. Also this won't help if there are no installed dependencies at all.

We then need to consider, in which scenarios we would expect deviations between `./libs/composer/vendor` and `./libs/composer/composer.lock`. I can think of two:
* Admin updated ILIAS via git and forgot to call `composer install --no-dev`.
* Admin updated dependencies on his own discretion because (e.g.) security, patches, by mistake, ...

The deviation will not appear if admins obey the installation-instructions, which should state to either use packaged releases (that we will need to supply) or use git and `composer install --no-dev`. So we have two todos:
* [ ] Update installation instruction to tell people how to use packaged releases or git and `composer install --no-dev`.
* [ ] Update release process to yield packaged releases including production dependencies.

There won't be any deviations between `composer.lock` and `.\vendor` in our supported use-cases then, so we do not need to inform admins about required updates, like we do with db-updates.

Finally, we could ask in which scenarios we would require the information if `composer.lock` and `.\vendor` are in sync. I could think of these:
* Admin ask himself if everything ok with the dependencies of her installation.
* Developer suspects something funky in the dependencies of an installation when investigating an issue.

I do not think that we need to pay special attention to the further. Admins that ask this question are already doing something interesting and should be able to answer it by themselves (most possibly by using git and/or composer). For the latter scenario, we need to consider how detailed the information for the developer need to be. IMO two information would be sufficient to determine if the correct dependencies are indeed installed:
* a hash of `composer.lock`
* a hash of the `./vendor`.

So I would propose a third todo:
* [x] Create a script located in `./libs/composer` that outputs a hash of `composer.lock` and a hash of `./vendor`.

Any opinions on that proposal, anyone?

**Comment by mjansenDatabay** [comment]
@klees Could you please fix the remaining conflicts? I will merge this afterwards.

As decided during our @ILIAS-eLearning/technical-board meeting, the thoughts/things you outlined above should be considered separately.

**Comment by chfsx** [comment]
@klees is it intended that the following files are still part of the repository?
I have `/libs/composer/vendor` in my .gitignore file now, but those two files still show changes after updating the classmap... 

`libs/composer/vendor/composer/autoload_classmap.php`
`libs/composer/vendor/composer/autoload_static.php`

**Comment by chfsx** [comment]
a `git rm -r libs/composer/vendor` did help

**Comment by KissDaniGH** [comment]
Hi @klees ,
I host a server for ILIAS and I didnt find a way to run any command (only FTP, phpmyadmin).
Is there any way to use only ftp to install ILIAS?
Bests, Daniel

**Comment by klees** [comment]
Hi @KissDaniGH,

you could well run `composer install --no-dev` at a local box and upload the result via ftp.

Feel free to contact me via e-mail (richard.klees@concepts-and-training.de) to discuss the options. Maybe we could derive some documentation for the core.

Best regards!

**Pull request merged**